messaging/internal/service/webhook.go

```diff
@@ -0,0 +1,254 @@
+		OutgoingURL:     params.OutgoingURL,
+	}
+
+	if !svc.perms.CanManageWebhooks(webhook) && !svc.perms.CanManageOwnWebhooks(webhook) {
+		return nil, errors.WithStack(ErrNoPermissions)
+	}
+
+	botUser := &systemTypes.User{
+		Username:      params.Username,
+		Name:          params.Username,
+		Handle:        params.Username,
+		Kind:          systemTypes.BotUser,
+		RelatedUserID: userID,
+	}
+
+	user, err := svc.users.CreateWithAvatar(botUser, params.Avatar)
+	if err != nil {
+		return nil, err
+	}
+	webhook.UserID = user.ID
+
+	wh, err := svc.webhook.Create(webhook)
+	if err != nil {
+		// cross service rollback (delete user)
+		svc.users.Delete(user.ID)
+		return nil, err
+	}
+	return wh, err
+}
+
+func (svc *webhook) Update(webhookID uint64, kind types.WebhookKind, channelID uint64, params types.WebhookRequest) (*types.Webhook, error) {
+	var userID = repository.Identity(svc.ctx)
+	webhook, err := svc.Get(webhookID)
+	if err != nil {
+		return nil, err
+	}
+
+	if !svc.perms.CanManageWebhooks(webhook) && !(webhook.OwnerUserID == userID && svc.perms.CanManageOwnWebhooks(webhook)) {
+		return nil, errors.WithStack(ErrNoPermissions)
+	}
+
+	botUser, err := svc.users.FindByID(webhook.UserID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error when looking for User ID %d", webhook.UserID)
+	}
+
+	if _, err := svc.users.UpdateWithAvatar(botUser, params.Avatar); err != nil {
+		return nil, err
+	}
+
+	webhook.Kind = kind
+	webhook.ChannelID = channelID
+	webhook.OutgoingTrigger = params.OutgoingTrigger
+	webhook.OutgoingURL = params.OutgoingURL
+
+	return svc.webhook.Update(webhook)
+}
+
+func (svc *webhook) Get(webhookID uint64) (*types.Webhook, error) {
+	return svc.webhook.Get(webhookID)
+}
+
+func (svc *webhook) Find(filter *types.WebhookFilter) (types.WebhookSet, error) {
+	return svc.webhook.Find(filter)
+}
+
+func (svc *webhook) Delete(webhookID uint64) error {
+	webhook, err := svc.Get(webhookID)
+	if err != nil {
+		return err
+	}
+	if svc.perms.CanManageWebhooks(webhook) {
+		return svc.webhook.Delete(webhookID)
+	}
+	var userID = repository.Identity(svc.ctx)
+	if webhook.OwnerUserID == userID && svc.perms.CanManageOwnWebhooks(webhook) {
+		return svc.webhook.Delete(webhookID)
+	}
+	return errors.WithStack(ErrNoPermissions)
+}
+
+func (svc *webhook) DeleteByToken(webhookID uint64, webhookToken string) error {
+	return svc.webhook.DeleteByToken(webhookID, webhookToken)
+}
+
+func (svc *webhook) Message(webhookID uint64, webhookToken string, username string, avatar io.Reader, message string) (interface{}, error) {
+	if webhook, err := svc.webhook.GetByToken(webhookID, webhookToken); err != nil {
+		return nil, err
+	} else {
+		msg := &types.Message{
+			Message: message,
+			Meta: &types.MessageMeta{
+				Username: username,
+			},
+		}
+		return svc.sendMessage(webhook, msg, avatar)
+	}
+}
+
+// Do executes an outgoing HTTP webhook request
+func (svc *webhook) Do(webhook *types.Webhook, message string) (*types.Message, error) {
```

Comment:
add some log points throughout this func (logging steps, latency etc...)

Reply:
I'm gonna open up this as a separate issue, since we don't have logging set up yet

Reply:
better to use what we have now (`log.Print...`).